### PR TITLE
[common-go] Avoid "bind: address already in use" during tests

### PR DESCRIPTION
**Pull request by geropl**
## Description

Avoid build errors like [this](https://github.com/gitpod-io/gitpod/actions/runs/5264242467/jobs/9515287095?pr=17831#step:14:4380). 

## Related Issue(s)

Fixes #

## How to test
See that tests are succeeding.

#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

**Review by akosyakov — APPROVED** [review]
let's try ❤️

File: components/common-go/baseserver/testing.go [hunk 1]
```diff
@@ -34,11 +46,11 @@ func NewForTests(t *testing.T, opts ...Option) *Server {
 	return srv
 }
 
-func MustUseRandomLocalAddress(t *testing.T) *ServerConfiguration {
+func MustUseUniqueLocalAddress(t *testing.T) *ServerConfiguration {
 	t.Helper()
 
 	return &ServerConfiguration{
-		Address: "localhost:0",
```
Comment (akosyakov) [inline components/common-go/baseserver/testing.go:41]: but 0 should pick the available one, no?

Comment (geropl) [reply components/common-go/baseserver/testing.go:41]: Yes. Googling hinted at either a) kernel bug or b) misconfigured network stack (tcp4 vs tcp6).
I didn't want to solve here :upside_down_face:

File: components/common-go/baseserver/testing.go [hunk 2]
```diff
@@ -14,6 +14,18 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+// go tests does execute tests concurrently (each test a goroutine)
+// As we ran into troubles with "localhost:0" (error: "failed to start HTTP server: listen tcp 127.0.0.1:0: bind: address already in use"),
+// which might be related to a kernel bug/config issues, we're rolling our own deduplication for now
+var lastPort = 24000
+
+func getTestAddr() string {
+	// No need to lock() as we do not execute in _parallel_ but _concurrently_
```
Comment (easyCZ) [inline components/common-go/baseserver/testing.go:23]: We do execute in parallel across packages

Comment (easyCZ) [reply components/common-go/baseserver/testing.go:23]: That's the default for go test

Comment (easyCZ) [reply components/common-go/baseserver/testing.go:23]: And the parallelism is further compounded by leeway doing the execution of multiple components at the same time

Comment (geropl) [reply components/common-go/baseserver/testing.go:23]: Awh, you're right, we using this all over the place :grimacing: 
Will do the retry then I guess.

File: components/common-go/baseserver/testing.go [hunk 3]
```diff
@@ -48,8 +49,24 @@ func StartServerForTests(t *testing.T, srv *Server) {
...
+			// NOTE: If you ask for a specific port (not ":0"), the test still fails
+			if strings.Contains(err.Error(), ":0: bind: address already in use") {
+				time.Sleep(time.Millisecond * 200)
+				continue
+			}
+
+			require.NoError(t, err)
+			return
+		}
+		t.Errorf("Cannot bind to %s after %d retries", srv.options.config.Services.HTTP.Address, retry)
```
Comment (easyCZ) [inline components/common-go/baseserver/testing.go:69]: This doesn't actually get called, because it already exits on line 66

Comment (easyCZ) [reply components/common-go/baseserver/testing.go:69]: Actually, it does, after 3 iterations. My bad

**Review by easyCZ — APPROVED** [review]

**Pull request merged**